---
title: "assignment week 06"
author: "Deni Hodzic"
output: html_document
date: '2022-04-01'
---

```{r}
suppressPackageStartupMessages({
  library(GenomicRanges)
  library(memes)
  library(Biostrings)
  library(AnnotationHub)
  library(MotifDb)
  library(universalmotif)
  library(PWMEnrich)
  library(ggplot2)
})
```

# Obtaining peak sequences

For the purpose of this example, I'll use the REST peaks found on chromosome 1 of mouse C2C12 differentiated myocytes:

```{r}
download.file("https://www.encodeproject.org/files/ENCFF368VWJ/@@download/ENCFF368VWJ.bed.gz", "REST_ENCFF368VWJ.bed.gz")
peaks <- rtracklayer::import("REST_ENCFF368VWJ.bed.gz", format="NarrowPeak")
seqlevelsStyle(peaks) <- "Ensembl"
peaks_chr1 <- peaks[seqnames(peaks)=="1"]
```


Since the motif of a transcription factor is typically found towards the center or summit of the peak, it is common practice to look for motifs around the center of peaks. 
How far around depends on the resolution of the data, i.e. how narrow are your peaks (can be between 50bp to 500bp around).
The first thing we do, then, is to extract the regions around the peaks, and then obtain the sequence from those regions:

```{r}
peak_centers <- resize(peaks_chr1, fix="center", width=100)
# we get the genome sequence:
ah <- AnnotationHub()
genome <- ah[["AH68356"]]
# we get the sequences corresponding to the peak centers:
peak_seqs <- memes::get_sequence(peak_centers, genome)
peak_seqs
```

# Motif scanning

Motif scanning aims at finding the _occurrences_ of a known motif in a set of sequences. As this methodologically is fairly simple, it doesn't really matter what method one uses, though results may differ because of how the thresholds are set.

## Getting the desired motif

```{r}
# we search for "REST" in the motif database
motifs <- query(MotifDb, "REST")
motifs
# there are several matching motifs:
names(motifs)
# we select one:
motif <- motifs[["Mmusculus-HOCOMOCOv10-REST_MOUSE.H10MO.A"]]
# we visualize it:
view_motifs(motif)
```

## Scanning a set of sequences using the findMotifInstances:

I'll first need the wrapper function:

```{r}
#' findMotifInstances
#' A wrapper around TFBSTools for scanning motif occurence, handling the coordinate conversion as `memes` does.
#' @param seqs A set of sequences, e.g. `DNAStringSet`, optionally with coordinate as names to enable conversion.
#' @param motif A motif, in any format recognized by `universalmotif`
#' @param keepMatchedSeq Logical; whether to keep the matched sequence.
#' @param ... Passed to `TFBSTools::searchSeq`; can for instance be used to set
#'   the number of threads to use, e.g. with `mc.cores=2`
#'
#' @return A `GRanges` object
findMotifInstances <- function(seqs, motif, keepMatchedSeq=FALSE, ...){
  motif <- universalmotif::convert_motifs(motif, "TFBSTools-PWMatrix")
  x <- suppressWarnings(TFBSTools::searchSeq(motif, subject=seqs))
  peaks <- strsplit(gsub("-",":",names(seqs)), ":")
  if(all(lengths(peaks)==3)){ # convert relative coordinates to absolute
    chrs <- sapply(peaks,FUN=function(x) x[1])
    offsets <- sapply(peaks,FUN=function(x) as.integer(x[2]))-1L
    i <- rep(seq_along(x),lengths(x))
    return(GRanges(chrs[i], 
                   IRanges(offsets[i]+as.integer(unlist(lapply(x, start))),
                           offsets[i]+as.integer(unlist(lapply(x, end)))),
                   strand=unlist(lapply(x, strand)),
                   score=as.numeric(unlist(lapply(x, FUN=function(x) score(x))))
                   ))
  }
  x <- as(x, "GRanges")
  score(x) <- x$absScore
  keepFields <- c("score","relScore")
  if(keepMatchedSeq) keepFields <- c(keepFields, "siteSeqs")
  mcols(x) <- mcols(x)[,intersect(colnames(mcols(x)), keepFields)]
  x
}
```


Then I can get the motif instances like this:

```{r}
moi2 <- findMotifInstances(peak_seqs, motif, mc.cores=2) # running with 2 threads
moi2
```

# Answering the questions
## Of all the peaks, what proportion contains a motif for the factor?

```{r}
# all the peaks:
length(peaks)
# peaks with motif
length(moi2)
# why does the following give me a smaller number than length(motif_instances)? shouldn't it be the same?
sum(overlapsAny(peaks, moi2))
percentage <- (sum(overlapsAny(peaks, moi2)))/length(peaks)*100
percentage
```
Result: Of all  the 3555 peaks, 90 ( 2.531646%) contain a motif.

## Of all instances of that motif in the genome, what proportion is bound by the factor (i.e. has a peak)?

```{r}
mmusculus <- import(genome, "2bit", which = as(seqinfo(genome), "GenomicRanges"))
motif_instances_genome <- findMotifInstances(mmusculus, motif, mc.cores=2)
length(motif_instances_genome)
```

```{r}
motif_with_peaks = overlapsAny(motif_instances_genome, peaks)
sum(motif_with_peaks)
percentage2 <- sum(motif_with_peaks)/length(motif_instances_genome)*100
percentage2
```

Result: Of all the 209697 motif instances, 2071 (0.9876155%) overlap a peak.
